File: 496-next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> stack = new Stack<>();
        HashMap<Integer,Integer> store = new HashMap<>();
        int[] ans = new int[nums1.length];
        int e = nums2.length - 1;
        while(e >= 0){
            if(stack.isEmpty()){
                stack.push(nums2[e]);
                store.put(nums2[e], -1);
            }else{
                while(!stack.isEmpty() && stack.peek() <= nums2[e]){
                    System.out.println(stack.peek());
                    stack.pop();
                }
                if(stack.isEmpty()){
                    stack.push(nums2[e]);
                    store.put(nums2[e], -1);
                }else{
                    
                    System.out.println("stack"+stack.peek()+"curr"+nums2[e]);
                    store.put(nums2[e], stack.peek());
                    stack.push(nums2[e]);
                }
            }
            e-=1;
        }
        for(int i = 0; i < nums1.length; i++){
            ans[i] = store.get(nums1[i]);
        }
        return ans;
    }
}

File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        store = {}
        ans = []
        for val in nums2[::-1]:
            if len(stack) == 0:
                stack.append(val)
                store[val] = -1
            else:
                while(stack and stack[-1] < val):
                    stack.pop()
                
                if stack:
                    store[val] = stack[-1]
                else:
                    store[val] = -1
                stack.append(val)
        
        for val in nums1:
            ans.append(store[val])

        return ans